src/SonarScanner.MSBuild.PreProcessor/JreCaching/IJreCache.cs

```diff
@@ -0,0 +1,26 @@
+﻿/*
+ * SonarScanner for .NET
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto: info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarScanner.MSBuild.PreProcessor.JreCaching;
+
+public interface IJreCache
+{
+    JreCacheResult IsJreCached(string sonarUserHome, JreDescriptor jreDescriptor);
```

Comment:
@gregory-paidis-sonarsource 
This service is supposed to be used by the server implementation like so:
* After the jre API endpoint is called, the WebServer calls the `IsJreCached` method with a `JreDescriptor` constructed from the returned JSON
* If `JreCacheResult` is `JreCacheMiss` then 
  * The `DownloadJRE` method of `IJreCache` will be called by WebServer: See below for the signature. This method will be defined and implemented in a follow-up PR.
  * The `Action<Task<Stream>> jreDownload` parameter will be provided by the WebServer and corresponds to something like [HttpContent.ReadAsStreamAsync](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpcontent.readasstreamasync?view=net-8.0).
  * `DownloadJRE` then takes this stream, and does all the required steps for the caching (storing on disk, checksum test, extraction

The idea is to decouple the cache handling from the WebServer API interaction. What do you think?

```cs
public interface IJreCache
{
    // ...
    JreCacheResult DownloadJRE(string sonarUserHome, JreDescriptor jreDescriptor, Action<Task<Stream>> jreDownload);
```

Reply:
It makes sense to me.
From what I understand, you want to isolate the not-identical-logic between SC and SQ in the `WebServer` derived implementations, and the `IJreCache` to contain all the identical logic only once (give-me-java part).